### PR TITLE
feat: support for the sort command[draft]

### PR DESCRIPTION
**Pull request by PokIsemaine**
* issues: https://github.com/apache/kvrocks/issues/1952

I'm trying to support the `sort` command and later the `sort_ro` command for kvrocks, this is my first time submitting a pull request for this project and I need some advice, so I set it to draft temporarily.

The help I need is as follows：
* There are three TODOs in the code and I don’t know how to deal with them. It seems that they are related to `ACL`, `Cluster`, `CLIENT_SCRIPT`, but I have not found relevant support for kvrocks at present. What should I do? The following is the relevant code in the redis implementation:
    * TODO1: https://github.com/redis/redis/blob/f4481e657f905074fa515701af3f695757817d88/src/sort.c#L237
    * TODO2: https://github.com/redis/redis/blob/f4481e657f905074fa515701af3f695757817d88/src/sort.c#L218
    * TODO3: https://github.com/redis/redis/blob/f4481e657f905074fa515701af3f695757817d88/src/sort.c#L296

* How to support `denyoom` and `movablekeys` flags, or whether kvrocks needs to support it?
```
127.0.0.1:6379> command info sort
1)  1) "sort"
    2) (integer) -2
    3) 1) write
       2) denyoom
       3) movablekeys
```
* Whether there is any need to modify or add to the existing code (especially exceptions), comments, and tests (especially boundary cases) in the draft?

If you can provide me with any advice I would be very grateful！
    

The information and code I referenced are as follows:
* [1] https://redis.io/commands/sort/
* [2] https://redis.io/commands/sort_ro/
* [3] https://github.com/redis/redis/blob/unstable/src/sort.c imitate its implementation
* [4] Execute `command docs sort` and `command info sort` to obtain the parameters, flags and other information of the sort command

### CODE REVIEW

**Comment by mapleFU** [comment]
> It seems that they are related to ACL, Cluster, CLIENT_SCRIPT, but I have not found relevant support for kvrocks at present. What should I do? 

I think we can leave them unimplemented first

> How to support denyoom and movablekeys flags

About oom currently it's a bit hard to manage the memory by custom allocator. Maybe we can separate a machine/process level global memory tracking module and check that. This can separate to an other patch or discuss it firstly

**Comment by PokIsemaine** [comment]
> > It seems that they are related to ACL, Cluster, CLIENT_SCRIPT, but I have not found relevant support for kvrocks at present. What should I do?
> 
> I think we can leave them unimplemented first
> 
> > How to support denyoom and movablekeys flags
> 
> About oom currently it's a bit hard to manage the memory by custom allocator. Maybe we can separate a machine/process level global memory tracking module and check that. This can separate to an other patch or discuss it firstly

Thanks!
So what do we need to do to modify these TODO comments? Delete them or keep them (perhaps with more formal comments)

**Review by jihuayu — COMMENTED** [review]
@PokIsemaine Thanks for your PR. Your code is amazingly clean. The test cases are also very detailed.
For these TODO comments, I think we can keep them.

The logic LGTM.  
But according to our conventions, this command should be placed in the `cmd_key.cc` file, and the core logic of the Execute function should be extracted and placed in the `redis::Database` class.

Below is a example:
https://github.com/apache/kvrocks/blob/26cbdb978b53c08216c6dfb5d45ed950ccde5936/src/commands/cmd_key.cc#L138-L149

File: src/commands/commander.h [hunk 1]
```diff
@@ -185,6 +187,10 @@ inline uint64_t ParseCommandFlags(const std::string &description, const std::str
       flags |= kCmdCluster;
     else if (flag == "no-dbsize-check")
       flags |= kCmdNoDBSizeCheck;
+    else if (flag == "deny-oom")
+      flags |= kCmdDenyOom;
+    else if (flag == "movable-keys")
+      flags |= kCmdMovableKeys;
```
Comment (PragmaTwice) [inline src/commands/commander.h]: This flag is meaningless in Kvrocks. Please remove it and refer to KeyRangeVecGen.

File: src/commands/commander.h [hunk 2]
```diff
@@ -185,6 +187,10 @@ inline uint64_t ParseCommandFlags(const std::string &description, const std::str
       flags |= kCmdCluster;
     else if (flag == "no-dbsize-check")
       flags |= kCmdNoDBSizeCheck;
+    else if (flag == "deny-oom")
+      flags |= kCmdDenyOom;
```
Comment (PragmaTwice) [inline src/commands/commander.h]: I think we don't plan to implement it in this PR, so please remove it.

File: src/commands/commander.h [hunk 3]
```diff
@@ -65,6 +65,8 @@ enum CommandFlags : uint64_t {
   kCmdROScript = 1ULL << 10,       // "ro-script" flag for read-only script commands
   kCmdCluster = 1ULL << 11,        // "cluster" flag
   kCmdNoDBSizeCheck = 1ULL << 12,  // "no-dbsize-check" flag
+  kCmdDenyOom = 1ULL << 13,        // "deny-oom" flag
+  kCmdMovableKeys = 1ULL << 14,    // "movable-keys" flag
```
Comment (PragmaTwice) [inline src/commands/commander.h]: ditto

File: src/commands/cmd_sort.cc [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+
+  std::string sortby_;                    // BY
+  bool dontsort_ = false;                 // DONT SORT
+  long offset_ = 0;                       // LIMIT OFFSET
+  long count_ = -1;                       // LIMIT COUNT
+  std::vector<std::string> getpatterns_;  // GET
+  bool desc_ = false;                     // ASC/DESC
+  bool alpha_ = false;                    // ALPHA
+  std::string storekey_;                  // STORE
+};
```
Comment (PragmaTwice) [inline src/commands/cmd_sort.cc]: The implementation of SORT should be in `redis_db.cc`. The parsing part can be in `cmd_key.cc`.

**Comment by PokIsemaine** [comment]
> @PokIsemaine Thanks for your PR. Your code is amazingly clean. The test cases are also very detailed. For these TODO comments, I think we can keep them.
> 
> The logic LGTM. But according to our conventions, this command should be placed in the `cmd_key.cc` file, and the core logic of the Execute function should be extracted and placed in the `redis::Database` class.
> 
> Below is a example:
> 
> https://github.com/apache/kvrocks/blob/26cbdb978b53c08216c6dfb5d45ed950ccde5936/src/commands/cmd_key.cc#L138-L149

Thanks for the advice, I'll probably do it this weekend

File: src/commands/commander.h (hunk 1 above)
Comment (PokIsemaine) [reply src/commands/commander.h]: Okay, I need to learn about KeyRangeVecGen.

File: src/commands/commander.h (hunk 2 above)
Comment (PokIsemaine) [reply src/commands/commander.h]: I agree with this, thanks for the heads up, I'll remove the flags later on.
